### PR TITLE
[MRG] MAINT: remove variables not needed to store

### LINKED ISSUE #134: Avoid storing any unnecessary variable after fit
Some algorithms currently store a reference to the data at fit time (having a `self.X_` attribute for instance). According to scikit-learn guidelines (https://scikit-learn.org/stable/developers/contributing.html#fitting), we should avoid this:
>Note that the model is fitted using X and y, but the object holds no reference to X and y. There are, however, some exceptions to this, as in the case of precomputed kernels where this data must be stored for use by the predict method.

If we really need all the data we could maybe store a copy of X but I don't think it is the case in metric-learn's algorithms

### PR DESCRIPTION
**Pull request by wdevazelhes**
Fixes #134

### CODE REVIEW

File: metric_learn/itml.py [hunk 1]
```diff
@@ -167,6 +181,17 @@ class ITML_Supervised(_BaseITML, TransformerMixin):
 
   Attributes
   ----------
+  bounds_ : `list` of two numbers
+      Bounds on similarity, aside slack variables, s.t. ``d(a, b) < pos`` for
+      all given pairs of similar points ``a`` and ``b``, and
+      ``d(c, d) > neg`` for all given pairs of dissimilar points ``c`` and
+      ``d``, with ``bounds=[pos, neg]``, and ``d`` the learned distance. If
+      not provided at initialization, these are the ones derived at train
+      time.
```
Comment (wdevazelhes) [inline metric_learn/itml.py]: I am not familiar with `ITML` but the above is based on what I understood of it. Feel free to tell me if I'm wrong

File: metric_learn/itml.py [hunk 2]
```diff
@@ -151,8 +161,12 @@ def fit(self, pairs, y, bounds=None):
...
     y: array-like, of shape (n_constraints,)
         Labels of constraints. Should be -1 for dissimilar pair, 1 for similar.
-    bounds : list (pos,neg) pairs, optional
-        bounds on similarity, s.t. d(X[a],X[b]) < pos and d(X[c],X[d]) > neg
+    bounds : `list` of two numbers
+        Bounds on similarity, aside slack variables, s.t. ``d(a, b) < pos`` for
+        all given pairs of similar points ``a`` and ``b``, and
+        ``d(c, d) > neg`` for all given pairs of dissimilar points ``c`` and
+        ``d``, with ``bounds=[pos, neg]``, and ``d`` the learned distance.
+        If not provided at initialization, these will be derived at train time.
```
Comment (wdevazelhes) [inline metric_learn/itml.py]: I also updated the description of the argument in `fit` (`bounds`), to match the one of `self.bounds_`, and to be more detailed. Same comment: feel free to tell me if I'm wrong

File: metric_learn/itml.py (hunk 1 above)
Comment (perimosocordiae) [reply metric_learn/itml.py]: As currently implemented, `bounds_` should be an ndarray-like of shape (2,).

We should probably update the code to accept list or tuple types for the bounds as well.

Comment (perimosocordiae) [reply metric_learn/itml.py]: "these are the ones" -> "these are"

File: metric_learn/itml.py [hunk 3]
```diff
@@ -167,6 +181,17 @@ class ITML_Supervised(_BaseITML, TransformerMixin):
...
+  bounds_ : `list` of two numbers
+      Bounds on similarity, aside slack variables, s.t. ``d(a, b) < pos`` for
+      all given pairs of similar points ``a`` and ``b``, and
+      ``d(c, d) > neg`` for all given pairs of dissimilar points ``c`` and
+      ``d``, with ``bounds=[pos, neg]``, and ``d`` the learned distance. If
+      not provided at initialization, these are the ones derived at train
+      time.
+
+  n_iter_ : `int`
+      The number of iterations the solver has ran.
```
Comment (perimosocordiae) [inline metric_learn/itml.py]: "has ran" -> "has run"

Comment (wdevazelhes) [reply metric_learn/itml.py]: >As currently implemented, bounds_ should be an ndarray-like of shape (2,).
>
>We should probably update the code to accept list or tuple types for the bounds as well.

That's right, I forgot about that. For now I'll just change the docstring, but I'll open an issue to get this thing straight

> "these are the ones" -> "these are"

Thanks, will do 


> "has ran" -> "has run"

That's right, thanks

File: metric_learn/itml.py [hunk 4]
```diff
@@ -181,16 +181,16 @@ class ITML_Supervised(_BaseITML, TransformerMixin):
 
   Attributes
   ----------
-  bounds_ : `list` of two numbers
+  bounds_ : array-like, shape=(2,)
       Bounds on similarity, aside slack variables, s.t. ``d(a, b) < pos`` for
       all given pairs of similar points ``a`` and ``b``, and
       ``d(c, d) > neg`` for all given pairs of dissimilar points ``c`` and
       ``d``, with ``bounds=[pos, neg]``, and ``d`` the learned distance. If
```
Comment (wdevazelhes) [inline metric_learn/itml.py]: I keep saying ``bounds=[pos, neg]`` here because it can be still be a list here, and I think this is more general than ``bounds=array([pos, neg])`` which would make think more of a numpy array only

File: metric_learn/itml.py [hunk 5]
```diff
@@ -134,6 +133,17 @@ class ITML(_BaseITML, _PairsClassifierMixin):
 
   Attributes
   ----------
+  bounds_ : array-like, shape=(2,)
+      Bounds on similarity, aside slack variables, s.t. ``d(a, b) < pos`` for
+      all given pairs of similar points ``a`` and ``b``, and
+      ``d(c, d) > neg`` for all given pairs of dissimilar points ``c`` and
+      ``d``, with ``bounds=[ pos, neg]``, and ``d`` the learned distance. If
+      not provided at initialization, these are derived at train
```
Comment (bellet) [inline metric_learn/itml.py]: maybe it would be nice to explain how it is done then: bounds_[0] and bounds_[1] are set to the 5th and 95th percentile of the pairwise distances among all points available at training?
for supervised version, of all points in the training data `X` ; for weakly supervised, of all points present in `pairs`

File: metric_learn/itml.py (hunk 2 above)
Comment (bellet) [inline metric_learn/itml.py]: again

File: metric_learn/itml.py [hunk 6]
```diff
@@ -167,6 +181,17 @@ class ITML_Supervised(_BaseITML, TransformerMixin):
 
   Attributes
   ----------
+  bounds_ : array-like, shape=(2,)
+      Bounds on similarity, aside slack variables, s.t. ``d(a, b) < pos`` for
+      all given pairs of similar points ``a`` and ``b``, and
+      ``d(c, d) > neg`` for all given pairs of dissimilar points ``c`` and
+      ``d``, with ``bounds=[pos, neg]``, and ``d`` the learned distance. If
+      not provided at initialization, these are derived at train
+      time.
```
Comment (bellet) [inline metric_learn/itml.py]: again

File: metric_learn/itml.py [hunk 7]
```diff
@@ -193,8 +218,12 @@ def __init__(self, gamma=1., max_iter=1000, convergence_threshold=1e-3,
...
     num_constraints: int, optional
         number of constraints to generate
-    bounds : list (pos,neg) pairs, optional
-        bounds on similarity, s.t. d(X[a],X[b]) < pos and d(X[c],X[d]) > neg
+    bounds : `list` of two numbers
+      Bounds on similarity, aside slack variables, s.t. ``d(a, b) < pos`` for
+      all given pairs of similar points ``a`` and ``b``, and ``d(c, d) > neg``
+      for all given pairs of dissimilar points ``c`` and ``d``, with
+      ``bounds=[pos, neg]``, and ``d`` the learned distance. If not provided at
+      initialization, these will be derived at train time.
```
Comment (bellet) [inline metric_learn/itml.py]: again

File: metric_learn/itml.py (hunk 5 above)
Comment (wdevazelhes) [reply metric_learn/itml.py]: Yes I agree it's more precise

File: metric_learn/itml.py (hunk 2 above)
Comment (wdevazelhes) [reply metric_learn/itml.py]: will do

File: metric_learn/itml.py (hunk 6 above)
Comment (wdevazelhes) [reply metric_learn/itml.py]: will do

File: metric_learn/itml.py (hunk 7 above)
Comment (wdevazelhes) [reply metric_learn/itml.py]: will do

**Comment by wdevazelhes** [comment]
Regarding `bounds`, I also noticed that for the supervised version it has to be given at initalization, and for the weakly supervised version it has to be put at fit. I think we should uniformize that. Since it is a data dependent parameter (indeed the default setting is deduced from the data), I think we should put them in fit. Since this is out of the scope of this PR and it should have a deprecation warning (which adds again code in this PR), I'll open another PR for that, and for this one I'll let the parameter where they are.

**Comment by wdevazelhes** [comment]
I adressed all the comments, so we should be good to merge this one

File: metric_learn/itml.py [hunk 8]
```diff
@@ -134,6 +133,18 @@ class ITML(_BaseITML, _PairsClassifierMixin):
 
   Attributes
   ----------
+  bounds_ : array-like, shape=(2,)
+      Bounds on similarity, aside slack variables, s.t. ``d(a, b) < pos`` for
+      all given pairs of similar points ``a`` and ``b``, and
+      ``d(c, d) > neg`` for all given pairs of dissimilar points ``c`` and
+      ``d``, with ``bounds=[ pos, neg]``, and ``d`` the learned distance. If
```
Comment (bellet) [inline metric_learn/itml.py]: My suggestion was to get rid of bounds=[ pos, neg] altogether and replace pos and neg in the inequalities by bounds_[0] and bounds_[1]

Comment (wdevazelhes) [reply metric_learn/itml.py]: Ah yes I agree it's even clearer, done
